nucliadb/src/nucliadb/writer/api/v1/knowledgebox.py

```diff
@@ -32,61 +38,101 @@
     KnowledgeBoxObjID,
     NucliaDBRoles,
 )
-from nucliadb_protos.knowledgebox_pb2 import (
-    DeleteKnowledgeBoxResponse,
-    KnowledgeBoxID,
-    KnowledgeBoxNew,
-    KnowledgeBoxResponseStatus,
-    KnowledgeBoxUpdate,
-    NewKnowledgeBoxResponse,
-    UpdateKnowledgeBoxResponse,
-)
+from nucliadb_protos import knowledgebox_pb2
 from nucliadb_utils.authentication import requires
-from nucliadb_utils.utilities import get_ingest
+from nucliadb_utils.settings import is_onprem_nucliadb
+
+
+def only_for_onprem(fun):
+    @wraps(fun)
+    async def endpoint_wrapper(*args, **kwargs):
+        if not is_onprem_nucliadb():
+            raise HTTPException(
+                status_code=403,
+                detail="This endpoint is only available for onprem NucliaDB",
+            )
+        return await fun(*args, **kwargs)
 
+    return endpoint_wrapper
 
+
+@only_for_onprem
 @api.post(
     f"/{KBS_PREFIX}",
     status_code=201,
     summary="Create Knowledge Box",
-    response_model=KnowledgeBoxObj,
     tags=["Knowledge Boxes"],
     openapi_extra={"x-hidden-operation": True},
 )
 @requires(NucliaDBRoles.MANAGER)
 @version(1)
-async def create_kb(request: Request, item: KnowledgeBoxConfig):
-    ingest = get_ingest()
-    requestpb = KnowledgeBoxNew()
-    requestpb = parse_create_kb_request(item)
-    kbobj: NewKnowledgeBoxResponse = await ingest.NewKnowledgeBox(requestpb)  # type: ignore
-    if item.slug != "":
-        slug = item.slug
+async def create_kb(request: Request, item: KnowledgeBoxConfig) -> KnowledgeBoxObj:
+    driver = get_driver()
+    rollback_learning_config = None
+
+    kbid = KnowledgeBox.new_unique_kbid()
+
+    # Onprem KBs have to call learning proxy to create it's own configuration.
+    if item.learning_configuration:
+        user_learning_config = item.learning_configuration
     else:
-        slug = kbobj.uuid
-    if kbobj.status == KnowledgeBoxResponseStatus.OK:
-        return KnowledgeBoxObj(uuid=kbobj.uuid, slug=slug)
-    elif kbobj.status == KnowledgeBoxResponseStatus.CONFLICT:
+        logger.warning(
+            "No learning configuration provided. Default will be used.",
+            extra={"kbid": kbid},
+        )
+        # learning will choose the default values
+        user_learning_config = {}
+
+    # we rely on learning to return the updated configuration with defaults and
+    # any other needed values (e.g. matryoshka settings if available)
+    learning_config = await learning_proxy.set_configuration(kbid, config=user_learning_config)
+
+    # if KB creation fails, we'll have to delete its learning config
+    async def _rollback_learning_config(kbid: str):
+        try:
```

Comment:
general comment that will make it easier to test:

WDYT about abstracting all the internal logic into another function and have only http-related logic in the endpoint (error handling and response serialization)?

Reply:
will make it more readable too

Reply:
Done